src/com/owncloud/android/lib/resources/files/ReadFileVersionsOperation.java

```diff
@@ -0,0 +1,154 @@
+            DavPropertyNameSet propSet = WebdavUtils.getFileVersionPropSet();
+
+            query = new PropFindMethod(uri, propSet, DavConstants.DEPTH_1);
+            int status = client.executeMethod(query);
+
+            // check and process response
+            boolean isSuccess = (status == HttpStatus.SC_MULTI_STATUS || status == HttpStatus.SC_OK);
+
+            if (isSuccess) {
+                // get data from remote folder
+                MultiStatus dataInServer = query.getResponseBodyAsMultiStatus();
+                readData(dataInServer, client);
+
+                // Result of the operation
+                result = new RemoteOperationResult(true, query);
+                // Add data to the result
+                if (result.isSuccess()) {
+                    result.setData(versions);
+                }
+            } else {
+                // synchronization failed
+                client.exhaustResponse(query.getResponseBodyAsStream());
+                result = new RemoteOperationResult(false, query);
+            }
+        } catch (Exception e) {
+            result = new RemoteOperationResult(e);
+        } finally {
+            if (query != null)
+                query.releaseConnection();  // let the connection available for other methods
+
+            if (result == null) {
+                result = new RemoteOperationResult(new Exception("unknown error"));
+                Log_OC.e(TAG, "Synchronized file with id " + fileId + ": failed");
+            } else {
+                if (result.isSuccess()) {
+                    Log_OC.i(TAG, "Synchronized file with id " + fileId + ": " + result.getLogMessage());
+                } else {
+                    if (result.isException()) {
+                        Log_OC.e(TAG, "Synchronized with id " + fileId + ": " + result.getLogMessage(),
+                                result.getException());
+                    } else {
+                        Log_OC.e(TAG, "Synchronized with id " + fileId + ": " + result.getLogMessage());
+                    }
+                }
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Read the data retrieved from the server about the file versions.
+     *
+     * @param remoteData Full response got from the server with the version data.
+     * @param client     Client instance to the remote server where the data were retrieved.
+     */
+    private void readData(MultiStatus remoteData, OwnCloudClient client) {
+        versions = new ArrayList<>();
+
+        // parse data from remote folder
+        WebdavEntry we;
+        String splitElement = client.getNewWebdavUri(false).getPath();
+
+        // loop to update every child
+        for (int i = 1; i < remoteData.getResponses().length; ++i) {
+            we = new WebdavEntry(remoteData.getResponses()[i], splitElement);
+            versions.add(new FileVersion(fileId, we));
```

Comment:
we as a variable isn't needed and could be dropped since it is needed only for the versions.add call